queries.js

```diff
@@ -0,0 +1,234 @@
+const fullDateRegex = /^\d\d\d\d-\d\d-\d\d$/;
+
+module.exports = (self, query) => {
+  return {
+    builders: {
+      upcoming: {
+        def: null,
+        async finalize() {
+          // Navigation by year, month or day should
+          // trump this filter allowing you to
+          // browse the past
+
+          if (query.get('year')) {
+            return;
+          }
+          if (query.get('month')) {
+            return;
+          }
+          if (query.get('day')) {
+            return;
+          }
+
+          const upcoming = query.get('upcoming');
+
+          if (upcoming === null) {
+            return;
+          }
+
+          if (upcoming) {
+            query.and({
+              end: { $gt: new Date() }
+            });
+          } else {
+            query.and({
+              end: { $lte: new Date() }
+            });
+          }
+        },
+        launder(value) {
+          return self.apos.launder.booleanOrNull(value);
+        },
+        choices() {
+          return [
+            { value: null, label: 'Both' },
+            { value: true, label: 'Upcoming' },
+            { value: false, label: 'Past' }
+          ];
+        }
+      },
+      // Filter by year, in YYYY-MM-DD format. The event must
+      // be taking place during that year (it might surround it).
+      // Use of this filter cancels the upcoming filter
+      year: {
+        def: null,
+        async finalize() {
+          const year = query.get('year');
+          if (!year) {
+            return;
+          }
+
+          query.and({
+            $and: [
+              { startDate: { $lte: year + '-12-31' } },
+              { startDate: { $gte: year + '-01-01' } }
+            ]
+          });
+        },
+        launder(value) {
+          return self.apos.launder.string(value);
+        },
+        async choices() {
+          const alldates = await query
+            .clone()
+            .upcoming(null)
+            .toDistinct('startDate');
+
+          const years = [{ value: null, label: 'All' }];
+          for (const eachdate of alldates) {
+            const year = eachdate.substr(0, 4);
+            if (!years.find(e => e.value === year)) {
+              years.push({ value: year, label: year });
+            }
+          }
+          years.sort().reverse();
+          return years;
+        }
+      },
+      // Filter by month, in YYYY-MM- format, using regex. The event must
+      // be taking place during that month (it might surround it).
+      // Use of this filter cancels the upcoming filter
+      month: {
+        def: null,
+        async finalize() {
```

Comment:
Currently doesn't do any actual async work (just modifies the query), so you can drop the keyword.

Reply:
Done